Add saga prototype.

drets opened the pull request:
```
modified:   framework/src/main/java/com/eventor/Eventor.java
modified:   framework/src/main/java/com/eventor/api/annotations/Saga.java
modified:   framework/src/main/java/com/eventor/internal/reflection/ClassProcessor.java
modified:   framework/src/test/groovy/com/eventor/EventorTest.groovy
new file:   framework/src/test/groovy/com/eventor/university/CourseRegistration.java
deleted:    framework/src/test/groovy/com/eventor/university/RegisterCourse.java
deleted:    framework/src/test/groovy/com/eventor/university/RegistreCourse.java
```

drets committed:
Initial saga prototype has been added.
	modified:   framework/src/main/java/com/eventor/Eventor.java
	modified:   framework/src/main/java/com/eventor/api/annotations/Saga.java
	modified:   framework/src/main/java/com/eventor/internal/reflection/ClassProcessor.java
	modified:   framework/src/test/groovy/com/eventor/EventorTest.groovy
	new file:   framework/src/test/groovy/com/eventor/university/CourseRegistration.java
	deleted:    framework/src/test/groovy/com/eventor/university/RegisterCourse.java
	deleted:    framework/src/test/groovy/com/eventor/university/RegistreCourse.java

st-kurilin commented on framework/src/main/java/com/eventor/Eventor.java:
```diff
@@ -3,10 +3,7 @@
 import com.eventor.api.*;
 import com.eventor.internal.Akka;
 import com.eventor.internal.Collections3;
-import com.eventor.internal.meta.Info;
```
Our policy is to collapse imports if there are more than 3 imports from 1 package.

st-kurilin commented on framework/src/test/groovy/com/eventor/EventorTest.groovy:
```diff
@@ -87,4 +87,19 @@ class EventorTest extends Specification {
     }
+
+    def "Saga should handle event"() {
+        setup:
```
Setup step extracted for all tests.

st-kurilin commented on framework/src/test/groovy/com/eventor/EventorTest.groovy:
```diff
@@ -87,4 +87,19 @@ class EventorTest extends Specification {
+        ch.registerCourse(courseId)
+        Thread.sleep(500)
+        then:
+        instanceCreator.getInstanceOf(CourseRegistration.class).getCourseId() == courseId
```
What do we test here?

st-kurilin commented on framework/src/test/groovy/com/eventor/university/CourseRegistration.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.concurrent.TimeUnit;
+
+@Saga
+public class CourseRegistration {
```
Let's come up with some more realistic example. Like Passing some exam with time limit.

st-kurilin replied:
Let's consider next example: 
Person can send StartFinalExam cmd and SubmitFinalExam cmd. There are FinalExamStarted evt and FinalExamGraduated evt.
Course handle StartFinalExam cmd and generate FinalExamStarted event. It also handle SubmitFinalExam cmd and generate FinalExamGraduated event.
FinalExamGraduating Saga starts on FinalExamStarted. It ends on FinalExamGraduated. If FinalExamStarted didn't come during 1 hour after FinalExamStarted it should generate SubmitFinalExam(0) event.

st-kurilin commented on framework/src/test/groovy/com/eventor/university/RegistreCourse.java:
```diff
@@ -1,4 +0,0 @@
-package com.eventor.university;
-
-public class RegistreCourse {
```
How is it differ from RegisterCourse?

drets committed:
Merge https://github.com/st-kurilin/eventor

Conflicts:
	framework/src/main/java/com/eventor/Eventor.java
	framework/src/main/java/com/eventor/internal/reflection/ClassProcessor.java
	framework/src/test/groovy/com/eventor/EventorTest.groovy

drets committed:
Saga prototype has been modified.
	modified:   framework/src/main/java/com/eventor/Eventor.java
	new file:   framework/src/main/java/com/eventor/api/Scheduler.java
	modified:   framework/src/main/java/com/eventor/internal/Akka.java
	modified:   framework/src/main/java/com/eventor/internal/ClassProcessor.java
	modified:   framework/src/test/groovy/com/eventor/EventorTest.groovy
	modified:   framework/src/test/groovy/com/eventor/university/Course.java
	new file:   framework/src/test/groovy/com/eventor/university/CourseExamStarted.java
	deleted:    framework/src/test/groovy/com/eventor/university/CourseRegistration.java
	new file:   framework/src/test/groovy/com/eventor/university/ExamGraduating.java
	modified:   framework/src/test/groovy/com/eventor/university/FailedOnCourseResultsSubmition.java
	modified:   framework/src/test/groovy/com/eventor/university/Grades.java
	new file:   framework/src/test/groovy/com/eventor/university/StartCourseExam.java

st-kurilin commented on framework/src/main/java/com/eventor/Eventor.java:
```diff
@@ -118,6 +123,28 @@ private void handleEventByAggregate(final MetaAggregate eachMetaAggregate, final
+    private void handleEventBySaga(final MetaSaga eachMetaSaga, final Object event) {
+        for (MetaHandler eachMetaHandler : eachMetaSaga.eventHandlers) {
+            if (eachMetaHandler.expected.equals(event.getClass())) {
+                Collection<?> commands = eachMetaHandler.execute(instanceCreator.getInstanceOf(eachMetaSaga.origClass), event);
```
We do have not only commands here.

st-kurilin commented on framework/src/main/java/com/eventor/Eventor.java:
```diff
@@ -118,6 +123,28 @@ private void handleEventByAggregate(final MetaAggregate eachMetaAggregate, final
+                for (Object eachCommand : commands) {
+                    if (eachCommand instanceof Finish) {
+                        scheduler.cancel();
+                        return;
```
return?

st-kurilin commented on framework/src/main/java/com/eventor/Eventor.java:
```diff
@@ -118,6 +123,28 @@ private void handleEventByAggregate(final MetaAggregate eachMetaAggregate, final
+                        return;
+                    }
+                    else if (eachCommand instanceof Timeout) {
+                        scheduler.scheduleOnce(Duration.create(((Timeout) eachCommand).duration, ((Timeout) eachCommand).unit),
```
Let's extract ((Timeout) eachCommand) to variable.

st-kurilin commented on framework/src/main/java/com/eventor/api/Scheduler.java:
```diff
@@ -0,0 +1,9 @@
+
+import scala.concurrent.duration.FiniteDuration;
+
+public interface Scheduler {
```
I guess it should not be part of com.eventor.api package.

st-kurilin commented on framework/src/main/java/com/eventor/api/annotations/Saga.java:
```diff
@@ -1,4 +1,11 @@
+@Target(ElementType.TYPE)
+@Inherited
 public @interface Saga {
+    Class<?> initBy() default Saga.class;
```
Do we need "default Saga.class" here?

st-kurilin commented on framework/src/main/java/com/eventor/internal/Akka.java:
```diff
@@ -48,7 +49,24 @@ public Invokable createBroadcaster(Iterable<Listener> listeners) {
+
+            @Override
+            public void scheduleOnce(FiniteDuration duration, Object args) {
+                cancellable = system.scheduler().scheduleOnce(duration, mainActor, args, system.dispatcher(), null);
```
What does null mean here?

st-kurilin commented on framework/src/test/groovy/com/eventor/university/CourseExamStarted.java:
```diff
@@ -0,0 +1,11 @@
+package com.eventor.university;
+
+public class CourseExamStarted {
+    String courseId;
```
let's make it public final

drets committed:
update
